r1cs/gadgets/std/src/sha256.rs

```diff
@@ -0,0 +1,392 @@
+    0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
+    0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
+    0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
+    0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
+    0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
+    0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
+    0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
+    0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
+];
+
+#[allow(clippy::unreadable_literal)]
+const IV: [u32; 8] = [
+    0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19,
+];
+
+pub fn sha256_block_no_padding<ConstraintF, CS>(
+    mut cs: CS,
+    input: &[Boolean],
+) -> Result<Vec<Boolean>, SynthesisError>
+where
+    ConstraintF: PrimeField,
+    CS: ConstraintSystem<ConstraintF>,
+{
+    assert_eq!(input.len(), 512);
+
+    Ok(
+        sha256_compression_function(&mut cs, &input, &get_sha256_iv())?
+            .into_iter()
+            .flat_map(|e| e.into_bits_be())
+            .collect(),
+    )
+}
+
+pub fn sha256<ConstraintF, CS>(mut cs: CS, input: &[Boolean]) -> Result<Vec<Boolean>, SynthesisError>
+where
+    ConstraintF: PrimeField,
+    CS: ConstraintSystem<ConstraintF>,
+{
+    assert!(input.len() % 8 == 0);
+
+    let mut padded = input.to_vec();
+    let plen = padded.len() as u64;
+    // append a single '1' bit
+    padded.push(Boolean::constant(true));
+    // append K '0' bits, where K is the minimum number >= 0 such that L + 1 + K + 64 is a multiple of 512
+    while (padded.len() + 64) % 512 != 0 {
+        padded.push(Boolean::constant(false));
+    }
+    // append L as a 64-bit big-endian integer, making the total post-processed length a multiple of 512 bits
+    for b in (0..64).rev().map(|i| (plen >> i) & 1 == 1) {
+        padded.push(Boolean::constant(b));
+    }
+    assert!(padded.len() % 512 == 0);
+
+    let mut cur = get_sha256_iv();
+    for (i, block) in padded.chunks(512).enumerate() {
+        cur = sha256_compression_function(cs.ns(|| format!("block {}", i)), block, &cur)?;
+    }
+
+    Ok(cur.into_iter().flat_map(|e| e.into_bits_be()).collect())
+}
+
+fn get_sha256_iv() -> Vec<UInt32> {
+    IV.iter().map(|&v| UInt32::constant(v)).collect()
+}
+
+fn sha256_compression_function<ConstraintF, CS>(
+    cs: CS,
+    input: &[Boolean],
+    current_hash_value: &[UInt32],
+) -> Result<Vec<UInt32>, SynthesisError>
+where
+    ConstraintF: PrimeField,
+    CS: ConstraintSystem<ConstraintF>,
+{
+    assert_eq!(input.len(), 512);
+    assert_eq!(current_hash_value.len(), 8);
+
+    let mut w = input
```

Comment:
Original code is:

```
    let mut w = input
        .chunks(32)
        .map(|e| UInt32::from_bits_be(e))
        .collect::<Vec<_>>();
```

Why don't you do it in the same way ? You force yourself to call UInt32::from_bits_le() each time you use w in the following lines

Reply:
I used this form because I need w to be a Vec<Vec<Boolean>> for the following operations in the sha256 code

Reply:
I don't get it sorry: later you re-transform the w chunks into UInt32, so what's the point of converting them into UInt32 then back into bits. Just to reverse them ? It's a bit dirty so I propose:

- Either we do it like Bellman and stick with w being a Vec<UInt32> (I would prefer);
- If you want to keep Vec<Vec<bool>> then let's reverse the bits normally using the reverse() function on Vec